File: UCFPT/211120/solutions/g/juggle.cpp
/*
  juggle.cpp
  JuggleFest, MCPC 2012 Problem G
  C++ solution by Michael Goldwasser
*/

#include <iostream>
#include <fstream>
using namespace std;

#define MAX_PERIOD 7
#define NUM_THROWS 20

ifstream fin("juggle.in");

int main() {
    while (true) {
        int P;
        fin >> P;
        if (P == 0) break;

        int pattern[MAX_PERIOD];
        for (int j=0; j < P; j++)
            fin >> pattern[j];

        string answer(20,'_');
        bool valid(true);
        int balls(0);
        for (int k=0; valid && k < NUM_THROWS; k++) {
            if (answer[k] == '_') {
                answer[k] = 'A' + balls++;     // start a new ball here
            }
            // forecast next occurrence of current ball
            int next = k + pattern[k % P];
            if (next < NUM_THROWS) {
                if (answer[next] == '_')
                    answer[next] = answer[k];
                else
                    valid = false;            // CRASH!!!!
            }
        }
        
        if (valid)
            cout << answer << endl;
        else
            cout << "CRASH" << endl;
    }
}

File: UCFPT/211120/solutions/g/juggle.java
/*
  juggle.java
  JuggleFest, MCPC 2012 Problem G
  Java solution by Andy Harrington
*/

import java.io.*;
import java.util.*;

public class juggle
{   
   static int p;
   static int[] seq;
   static int MAX=20;
   
   public static void main(String[] args) throws Exception {
      String file = (args.length > 0) ? args[0] : "juggle.in";
      Scanner in = new Scanner(new File(file));
      p  = in.nextInt();
      while (p  > 0) {
        seq = new int[p]; 
        for (int i = 0; i < p; i++)  
           seq[i] = in.nextInt();
        System.out.println(solve());
        p = in.nextInt();
      }
   }

   static String solve() {
      char[] toss = new char[MAX];
      char ch = 'A';
      for (int i = 0; i < MAX; i++) {
         if (toss[i] == 0)
            toss[i] = ch++;
         int hi = i + seq[i % p];
         if (hi < 20)
            if ( toss[hi] != 0)
               return "CRASH";
            else
               toss[hi] = toss[i];    
      }
      return new String(toss);
   }
}
